Provide link to documentation when test framework dependencies are missing (#33976)
gradle/gradle: platforms/documentation/docs/src/docs/userguide/platforms/jvm/java_testing.adoc

For information on how to determine code coverage for tests in additional source sets, see the <<jacoco_plugin#jacoco_plugin,JaCoCo Plugin>> and the <<jacoco_report_aggregation_plugin#jacoco_report_aggregation_plugin, JaCoCo Report Aggregation Plugin>> chapters.

[[sec:java_testing_troubleshooting]]

> == Troubleshooting common problems

Author: @lkasso could you take a look at this?

=== Tests are not detected or executed

If no tests are detected, Gradle will skip the `Test` task entirely.

> There could be a few causes:

Reviewer: ```suggestion
This can occur for several reasons:
```

>  * No test sources exist. This is normal behavior.

Reviewer: ```suggestion
 * No test sources exist. This is expected behavior for empty test suites.
```

>  * Test sources exist, but they were not compiled. The sources may be in the wrong directory or the source set may not be configured correctly.

Reviewer: ```suggestion
 * Test sources exist but are not compiled. This may happen if:
 ** The source files are located in the wrong directory.
 ** The source set is not configured correctly.
```

>  * No test classes are configured on the `Test` task. By default, `Test` tasks do not assume any particular path for test classes. Only the built-in task named `test` is automatically configured to use the `test` source set.

Reviewer: ```suggestion
 * No test classes are configured on the `Test` task. By default, only the built-in `Test` task is automatically configured to use the test source set.
```

>    If you have a custom `Test` task, you need to configure the `testClassesDirs` and `classpath` properties appropriately.

Reviewer: ```suggestion
   If you're using a custom `Test` task, you must manually configure both `testClassesDirs` and `classpath`.
```

Test dependencies need to be declared on the appropriate configuration.
The <<jvm_test_suite_plugin.adoc#jvm_test_suite_plugin,JVM Test Suite>> plugin automatically configures all necessary dependencies, but if you are manually configuring a `Test` task, you need to ensure that all required dependencies are declared on the appropriate configuration.

For JUnit4-based tests, you need to declare a dependency on a single JUnit 4.x library.
For TestNG-based tests, you need to declare a dependency on a single TestNG library.
For JUnit Jupiter-based tests, you need to declare dependencies on the JUnit Jupiter and JUnit platform libraries.

Before Gradle 9.0.0, Gradle <<upgrading_version_8#test_framework_implementation_dependencies,leaked internal dependencies>> into the test runtime classpath which may have masked missing dependencies.

=== Test process fails with unusual exit code

The test process can exit with an unusual exit code (non-zero).

This can be caused by a variety of issues, such as:

 * A test calls `System.exit()` directly.
 * Too many processes are running which causes the OS to kill the test process.
 * The test process runs out of memory.

> 

Author: I had the same thought when I wrote this, but I don't have good actionable advice.
